### ISSUE TITLE
Elided elements attributes cause errors even with `--iree-util-zero-fill-elided-attrs`

### ISSUE THREAD

**Issue by bjacob** [iree-org/iree#14415]
I rely on `--mlir-elide-elementsattrs-if-larger=10` as a first step in reducing large testcases.

Let me know if I'm missing something, but I thought that the resulting bytecode modules should be accepted by `iree-compile --iree-util-zero-fill-elided-attrs`. They're usually not at the moment; this issue seems to happen so consistently over the past few cases I've deal with, both f32 and i8 models, very different from each other.

Just as one instance, taking the testcase from https://github.com/openxla/iree/issues/14414 , doing

```
tools/iree-opt --mlir-elide-elementsattrs-if-larger=10 /tmp/stablehlo.mlirbc > /tmp/stablehlo.elided.mlir
```

then running the repro `iree-compile` command line with `--iree-util-zero-fill-elided-attrs` added:

```
$ tools/iree-compile --output-format=vm-bytecode     --mlir-print-op-on-diagnostic=false     --iree-hal-target-backends=llvm-cpu    --iree-input-type=stablehlo     --iree-llvmcpu-target-triple=x86_64-unknown-linux-gnu     --iree-llvmcpu-target-cpu=cascadelake    --iree-flow-enable-data-tiling     --iree-llvmcpu-enable-microkernels  --iree-util-zero-fill-elided-attrs   /tmp/stablehlo.elided.mlir -o /tmp/module.vmfb
```

Fails like this:

```
iree-compile: iree/third_party/llvm-project/llvm/include/llvm/Support/Casting.h:566: decltype(auto) llvm::cast(const From &) [To = mlir::DenseElementsAttr, From = mlir::ElementsAttr]: Assertion `isa<To>(Val) && "cast<Ty>() argument of incompatible type!"' failed.
```

My workaround is to rewrite elided attribute as `dense<0>` or `dense<0.0>` depending on whether the element type is floating point. For an all-float model like the one above, I do:

```
tools/iree-opt --mlir-elide-elementsattrs-if-larger=10 /tmp/stablehlo.mlirbc | sed 's/dense_resource<__elided__>/dense<0.0>/g' > /tmp/stablehlo.elided.mlir
```

**Comment by benvanik** [comment]
Probably just some bitrot from upstream changes - should be easy to fix.

**Comment by benvanik** [comment]
Added a test for serialization and it passes - was really confused - turns out with your particular repro isn't related to iree-util-zero-fill-elided-attrs and instead is a stablehlo -> linalg pattern that's dying. A stack trace would have shown where the assert was firing:

```
 #0 0x00007ff6d22ea9ec HandleAbort D:\Dev\iree\third_party\llvm-project\llvm\lib\Support\Windows\Signals.inc:419:0
 #1 0x00007fffa1a090ed (C:\WINDOWS\SYSTEM32\ucrtbased.dll+0xa90ed)
 #2 0x00007fffa1a0ae49 (C:\WINDOWS\SYSTEM32\ucrtbased.dll+0xaae49)
 #3 0x00007fffa1a11345 (C:\WINDOWS\SYSTEM32\ucrtbased.dll+0xb1345)
 #4 0x00007fffa1a10bd7 (C:\WINDOWS\SYSTEM32\ucrtbased.dll+0xb0bd7)
 #5 0x00007fffa1a0eba1 (C:\WINDOWS\SYSTEM32\ucrtbased.dll+0xaeba1)
 #6 0x00007fffa1a118af (C:\WINDOWS\SYSTEM32\ucrtbased.dll+0xb18af)
 #7 0x00007ff6c1603f36 D:\Dev\iree\third_party\llvm-project\llvm\include\llvm\Support\Casting.h:566:0
 #8 0x00007ff6c15c4d75 mlir::iree_compiler::stablehlo::`anonymous namespace'::ConstConverterTensor::matchAndRewrite D:\Dev\iree\compiler\src\iree\compiler\InputConversion\StableHLO\StableHLOToLinalg.cpp:1512:0
 #9 0x00007ff6c167d886 mlir::OpConversionPattern<class mlir::stablehlo::ConstantOp>::matchAndRewrite(class mlir::Operation *, class llvm::ArrayRef<class mlir::Value>, class mlir::ConversionPatternRewriter &) const D:\Dev\iree\third_party\llvm-project\mlir\include\mlir\Transforms\DialectConversion.h:539:0
#10 0x00007ff6cae6f835 mlir::ConversionPattern::matchAndRewrite(class mlir::Operation *, class mlir::PatternRewriter &) const D:\Dev\iree\third_party\llvm-project\mlir\lib\Transforms\Utils\DialectConversion.cpp:1664:0
#11 0x00007ff6caf21d45 `mlir::PatternApplicator::matchAndRewrite'::`8'::<lambda_1>::operator() D:\Dev\iree\third_party\llvm-project\mlir\lib\Rewrite\PatternApplicator.cpp:206:0
```

Bad cast is here:
https://github.com/google/iree/blob/1b139661974670c59701f0c6db3a7d1ebca1b1fe/compiler/src/iree/compiler/InputConversion/StableHLO/StableHLOToLinalg.cpp/#L1510

I'll fix the pattern to handle resource attrs.

/cc @kuhar

**Comment by kuhar** [comment]
Seems like an incorrect cast in the code we carried over from mlir-hlo. I can see a few other issues in the patterns in the same file.

TL;DR: rewrite patterns should not assume exact attribute type or exact type returned by the type converter.

**Comment by benvanik** [comment]
Ah yes, that'd make sense :)
![image](https://github.com/openxla/iree/assets/75337/e4e73d90-67b0-4f74-954f-38658e6e6d61)

I've got this one pattern fixed but as you mention there are probably other things lurking.
